### ISSUE TITLE
TransportSearchAction/CanMatchPreFilterSearchPhase may block transport_worker

### ISSUE THREAD

**Issue by idegtiarenko** [elastic/elasticsearch#109527]
### Elasticsearch Version

8.13.2

### Java Version

_bundled_

### OS Version

n/a

### Problem Description

TransportSearchAction/CanMatchPreFilterSearchPhase might block the `transport_worker` for significant amount of time.
This was detected with a search that matched 50k shards on a cluster with 80k total shards.

Please see attached hot threads outputs:
```
   100.8% [cpu=100.8%, idle=0.0%] (504.2ms out of 500ms) cpu usage by thread 'elasticsearch[instance-0000000816][transport_worker][T#31]'
     2/10 snapshots sharing following 165 elements
       java.base@21.0.2/java.util.Comparators$NaturalOrderComparator.compare(Comparators.java:52)
       java.base@21.0.2/java.util.Comparators$NaturalOrderComparator.compare(Comparators.java:47)
       app/org.apache.lucene.core@9.10.0/org.apache.lucene.util.CollectionUtil$ListTimSorter.compare(CollectionUtil.java:136)
       app/org.apache.lucene.core@9.10.0/org.apache.lucene.util.TimSorter.nextRun(TimSorter.java:118)
       app/org.apache.lucene.core@9.10.0/org.apache.lucene.util.TimSorter.sort(TimSorter.java:204)
       app/org.apache.lucene.core@9.10.0/org.apache.lucene.util.CollectionUtil.timSort(CollectionUtil.java:186)
       app/org.apache.lucene.core@9.10.0/org.apache.lucene.util.CollectionUtil.timSort(CollectionUtil.java:200)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.CanMatchPreFilterSearchPhase.<init>(CanMatchPreFilterSearchPhase.java:118)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction$AsyncSearchActionProvider.newSearchPhase(TransportSearchAction.java:1297)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeSearch(TransportSearchAction.java:1153)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeLocalSearch(TransportSearchAction.java:914)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.lambda$executeRequest$8(TransportSearchAction.java:342)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction$$Lambda/0x000000080290dc20.accept(Unknown Source)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:109)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:77)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeRequest(TransportSearchAction.java:455)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:309)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:113)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:96)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.ActionFilter$Simple.apply(ActionFilter.java:53)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:93)
   ...
        2/10 snapshots sharing following 158 elements
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.CanMatchPreFilterSearchPhase.<init>(CanMatchPreFilterSearchPhase.java:119)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction$AsyncSearchActionProvider.newSearchPhase(TransportSearchAction.java:1297)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeSearch(TransportSearchAction.java:1153)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeLocalSearch(TransportSearchAction.java:914)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.lambda$executeRequest$8(TransportSearchAction.java:342)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction$$Lambda/0x000000080290dc20.accept(Unknown Source)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.ActionListenerImplementations$ResponseWrappingActionListener.onResponse(ActionListenerImplementations.java:245)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:109)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.index.query.Rewriteable.rewriteAndFetch(Rewriteable.java:77)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.executeRequest(TransportSearchAction.java:455)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:309)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:113)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:96)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.ActionFilter$Simple.apply(ActionFilter.java:53)
       app/org.elasticsearch.server@8.13.2/org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:93)
...
```

### Steps to Reproduce

It is required to run a search query against the pattern that matches significant amount of shards.

**Comment by original-brownbear** [comment]
Fix for this already in progress in #109497 :)

**Comment by benwtrent** [comment]
@original-brownbear your fix references "freeing store references" where I think this issue is simply because the `CanMatchPreFilterSearchPhase::new` runs `timsort` on the shards and creates a large hashmap of all the shards being matched.

**Comment by original-brownbear** [comment]
Right I should've been clearer sorry :) I think depending on how my fix goes, it will probably cover this unless we go for a tricky fix for the closing issue. But it seems just forking more to search_worker for this action is even more interesting with this bug? Let's see :)

**Comment by benwtrent** [comment]
Here is an example as well: https://github.com/elastic/elasticsearch/issues/67031

We really shouldn't block the transport thread in scenarios like this bug (maybe we don't anylonger).

**Comment by original-brownbear** [comment]
I think this is far better now, I tested this with 50k shards and could not reproduce a macroscopic block any longer after merging #110860  and #110909 (could reproduce second scale blocking without those fixes).
Obviously, these two do not solve the problem for any shard count, but we have a bunch of O(N) scalability pieces on the transport thread and with the fixes applied this one isn't special in any way now.
=> closing here for now. If we want better scalability here we should look into solving those O(N) scalability issues. As we discussed, there is no reason to have them in theory and practically we'll incur a latency hit from any forking solution we apply to this problem. An alternative might be to have a yielding solution where we enqueue a task on the same transport thread, but again this should have a fix in architecture not in code if we want to scale way past O(10k) shards.
-> for now this is good enough I believe for the scales we support
